backend/uasmartsignage/src/main/java/deti/uas/uasmartsignage/Services/TemplateGroupService.java

```diff
@@ -130,6 +157,92 @@ public TemplateGroup updateTemplateGroup(Long id, TemplateGroup templateGroup) {
+            
+            // get a base html to add the widgets to
+            File baseFile = ResourceUtils.getFile("classpath:" + filePath);
+            Document doc = Jsoup.parse(baseFile, "UTF-8");
+
+            for (TemplateWidget widget : widgets) {
+                // fill in the variables inside the widgets
+                // using the values inside "contents"
+                String widgetHTML = loadWidget(widget, contents, monitorWidth, monitorHeight);
+                
+                if(widgetHTML == null){
+                    continue;
+                }   
+
+                // add the html of the widget to the main body
+                List<Element> widgetElements = Jsoup.parseBodyFragment(widgetHTML).body().children();
+                for(Element child : widgetElements)
+                    doc.body().appendChild(child);
+            }
+
+            return doc.html();
+
+        } catch (FileNotFoundException e) {
+            logger.error(String.format("Could not find file 'resources/%s'", filePath));
+            return null;
+        } catch (IOException e) {
+            logger.error(String.format("Jsoup could not parse the file 'resources/%s'", filePath));
+            return null;
+        }
+
+    }
+
+    /**
+     * Takes a widget and fills in the [[spaces]] with the Contents
+     * 
+     * @param widget The Widget that we want to complete
+     * @param contents A Map that contains the ids of Contents and the respective variables to fill in the widgets
+     * @param monitorWidth The screen width to be considered when calculating widget sizes
+     * @param monitorHeight The screen height to be considered when calculating widget sizes
+     * @return The created HTML in a String, or {@code null} if creation fails.
+     */
+    private String loadWidget(TemplateWidget widget, Map<Integer, String> contents, int monitorWidth, int monitorHeight) {
+
+        try {
+            File widgetFile = ResourceUtils.getFile("classpath:" + widget.getWidget().getPath());
+            String widgetHTML = new String(Files.readAllBytes(Paths.get(widgetFile.toURI())));
+
+            widgetHTML = widgetHTML
+                    .replace("[[top]]", String.valueOf(monitorHeight * widget.getTop() / 100))
+                    .replace("[[left]]", String.valueOf(monitorWidth * widget.getLeftPosition() / 100))
+                    .replace("[[width]]", String.valueOf(monitorWidth * widget.getWidth() / 100))
+                    .replace("[[height]]", String.valueOf(monitorHeight * widget.getHeight() / 100));
+
+            // go through each of the contents
+            // and fill in the variables
+            for(int contentID : contents.keySet()){
+                String value = contents.get(contentID);
+                Content content = contentService.getContentById((long)contentID);
+                widgetHTML = widgetHTML.replace("[[" + content.getName() + "]]", value);
+            }
+
+            return widgetHTML;
```

Comment:
Same as the previous one, having a log could be useful to identify what was loaded.